vizro-core/src/vizro/actions/_action_manager/_action_manager.py

```diff
@@ -0,0 +1,40 @@
+"""The action manager handles creation of all required actions callbacks."""
+
+from vizro.actions._action_manager._build_action_loop_callbacks import _build_action_loop_callbacks
+from vizro.actions._action_manager._get_action_loop_components import _get_action_loop_components
+from vizro.managers import model_manager
+from vizro.models import Action
+
+
+class ActionManager:
+    @classmethod
+    def build(cls):
+        """Builds callbacks for the action loop and for each Action in the Dashboard and returns their components.
+
+        Returns:
+            All required components for the action loop and for each Action in the Dashboard.
+        """
+        return cls._build_action_loop() + cls._build_action_models()
+
+    @staticmethod
+    def _build_action_loop():
+        """Builds callbacks for the action loop and gets all required components.
+
+        Returns:
+            All required components for the action loop.
+        """
+        _build_action_loop_callbacks()
+        return _get_action_loop_components()
+
+    @staticmethod
+    def _build_action_models():
+        """Builds a callback for each Action model in the Dashboard and gets all required components.
+
+        Returns:
+            All required components for each action model in the Dashboard.
+        """
+        return [
+            action_component
+            for _, action in model_manager._items_with_type(Action)
+            for action_component in action.build()
+        ]
```

Comment:
What's the difference in getting the action components in _build_action_models and _build_action_loop? Both seem to return "all required action components"? Can we provide a more distinct description in the docstrings?

Reply:
`_build_action_loop` builds callbacks and returns required components for smooth work of the `action loop`.
Otherwise, `_build_actions_models` builds callbacks and returns required components for each defined `Action model` in the system.

Reply:
I still don't get it 😄  What does smooth work of the action loop mean? Can you give actual output examples? 
What are the components returned by `_build_action_loop` and the ones by` _build_action_models`? Is one of them just returning dcc.Stores? While the other returns actual dash components such as dcc.Download?

Reply:
Looking the architecture given below:
 - "trigger_to_global_store",
 - "gateway",
 - "update_remaining_actions",
 - "executor",
 - "after_action",
 - and "clientside_callback" 

are callbacks created inside the `_build_action_loop_callbacks` function. These callbacks implements "action loop" mechanism that enables actions to be executed one by one in a sequence. 

---

`_get_action_loop_components` function returns all required components so that the "action loop" can work. Here are some examples of these components (and there are many more):
```
        dcc.Store(id="empty_input_store"),
        dcc.Store(id="empty_output_store"),
        dcc.Store(id="action_finished"),
        dcc.Store(id="set_remaining"),
        dcc.Store(id="remaining_actions"),
        html.Div(id="cycle_breaker_div", style={"display": "hidden"}),
        dcc.Store(id="cycle_breaker_empty_output_store"),
```

These two functions are wrapped into the one function called: `_build_action_loop`. 

---

Otherwise, `_build_actions_models` triggers `Action.build` method for each action model in the system.
Looking the architecture given below:
 - "Action_1",
 - "Action_2",
 - and "Action_3"

are callbacks created by Action.build method (each Action.build method creates one of the listed callbacks ("Action_1", "Action_2", "Action_3",)).

---

Following the pattern like the Action.build method creates callbacks and returns some components to be displayed, `_build_action_loop` and `_build_actions_models` do the same. 


<img width="1382" alt="image" src="https://github.com/mckinsey/vizro/assets/108530920/99583c0f-cf8a-4f73-ab8e-4377160f13e8">

Reply:
Got it, thanks! I went into the methods and looked at the output as well - correct me here if I am wrong:

The `_build_actions_models` return actual output components (e.g. dcc.Download) which are targeted by the action functions provided to the models e.g. Button.

The  `_build_action_loop` returns hidden components (such as dcc.Store/html.Div) required to store relevant data for the action loop engine to work.

This is what I wanted to know, because in the docstring description it's not clear what is meant with "returning action components". I would make a clear distinction in the docstrings and return type hints, such that it's easier to follow!

If the return of these are always components, shall we re-name to `_get_action_loop_components` and `_get_action_model_components`?

Reply:
You're right, `_build_action_loop` and `_build_actions_models` do what you said. But, they also create callbacks.  The `_build_action_loop` also creates callbacks for the `action loop` and the `_build_actions_models` also creates callbacks for `each Action model` in the Dashboard. 

So, why I didn't separate `ActionManager._create_app_callbacks` into something like:
 -  `ActionManager._create_app_callbacks`,
 -  `ActionManager._get_actions_components`
?  

The reason is that I want to follow our way of managing models. Every model has a `build` method which actually:
 - creates a callback and
 - returns needed components
(e.g. Slider, RangeSlider, Action and so on)

So, I wanted to follow the same pattern in the ActionManager too, by "exposing" only the `build` (or some other name like `_create_app_callbacks`) method that actually do the same: 
 - creates needed callbacks and
 - returns needed components.   

Hence, I wouldn't change methods names. 

--- 

I will update docstrings, thanks for an advice.

---

Also, I don't think that return type can bring any useful info. Any of our model.build methods don't have return type, so Action.build neither. It could seem confusing because it would be something like (List[Union[[dcc.Download, html.Div, dcc.Store]]]).    

Also, I don't think that return type can yield useful information. None of our model.build methods have a return type, and neither does Action.build. Also, it might seem confusing because it would be something like `List[Union[dcc.Download, html.Div, dcc.Store]]` (actually it should be `List[Union[any_dcc_object, any_html_object]]`).

Reply:
- On the first point, I agree. If it creates the callbacks and gets the components, then let's leave as is
- Perfect, thanks for updating the docstrings!
- On the third point, I kind of disagree, because it would have helped me at least to understand what the different components being returned are. One of the reasons why we don't have it in our model.build is because it's obvious from the code what is being returned. Remember in the build method we e.g. directly `return html.Div(...)`.

Here, I had to run the functions and take a look at the output to understand that these two methods return different "action components". Remember that someone not familiar with the action code should optimally understand what's happening in the action loop. So while it might be obvious for you right now, it might help others by providing more detailed docs and type hints :) 

Isn't the return of this:
_get_action_loop_components -> List[Union[html.Div, dcc.Store]]
_build_action_models -> List[dcc.Download] (at least right now, this will probably be extended further)

Reply:
You're right, I see. Let me implement it 😄.

Reply:
If the return type hints lead to lots of mypy issues, that are difficult to resolve, I would also be fine with adding the expected return types in the docstrings at least. So if option 1 is difficult to implement/maintain, that would also be an alternative for me :)

Reply:
I tried with the return type and as you guessed, I had a lot of trouble with mypy (), so I described it in docstrings.